# Практическая работа №3: Практическое применение библиотеки символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики.

Выполнил студент гр. 9383 Камзолов Никита.

## Постановка задачи

1. Решить свой вариант диофантова уравнения.
2. Имея открытую часть ключа RSA найти соответствующую закрытую часть и расшифровать кодированное слово.
3. Записать в формате ДНФ и КНФ булеву функцию, заданную своим аналитическим выражением.

## Выполнение работы

### 1. Решение диофантова уравнения

$2125x - 2737y = 34$

In [1]:
from sympy.solvers.diophantine.diophantine import diop_solve
from sympy import symbols

In [2]:
x, y = symbols("x, y", integer=True)
print("Множество всех решений:", diop_solve(2125*x-2737*y-34))

Множество всех решений: (161*t_0 + 152, 125*t_0 + 118)


Частное решение: $(152, 118)$

### 2. RSA-дешифрование

Русское слово из четырех букв закодировано при помощи алгоритма RSA открытым ключом (e = 5,m = 39).
Шифрованное сообщение имеет вид    (22; 14; 37; 3). Подберите закрытую часть ключа и прочитайте исходное слово.
Буквам русского алфавита соответсвуют числа в диапазоне от 2 до 32 (исключены буквы “ё” и “ъ”).

In [3]:
from sympy.crypto.crypto import decipher_rsa
from sympy.crypto.crypto import rsa_private_key

m = 39
e = 5

#Разложим m на простейшие
p = 3
q = 13
print(m == p*q)

True


In [4]:
private_key = rsa_private_key(p, q, e)

encrypted_word = [22, 14, 37, 3]
word = []
for value in encrypted_word:
    word.append(decipher_rsa(value, private_key))
print(word)

[16, 14, 7, 9]


Зашифрованное слово: ОМЕЗ

### 3. Запись в формате ДНФ и КНФ булевой функции, заданной аналитическим выражением.
$f(x, y, z) = ((xy)\vee (z\vee y))\ xor\ ((x\ xor\ z)y)$

In [5]:
from sympy.logic.boolalg import to_cnf, to_dnf, Xor
from sympy.abc import X, Y, Z

In [6]:
f = (Xor(((X & Y) | (Z | Y)), Xor(X,Z) & Y))

#### Конъюктивная нормальная форма:

In [7]:
to_cnf(f, simplify = True)

(Y | Z) & (Z | ~X) & (X | ~Y | ~Z)

#### Дизъюнктивная нормальная форма:

In [8]:
to_dnf(f, simplify = True)

(X & Z) | (Z & ~Y) | (Y & ~X & ~Z)

### Вывод.

Практически применена библиотека символьных вычислений Python (SymPy) для решения задач дискретной математики и математической логики.